From 5f4e161089c3d0882fb563158b0f115278dd9524 Mon Sep 17 00:00:00 2001
From: Alexandru Firuti <alexandru.firuti@nxp.com>
Date: Wed, 15 Oct 2025 15:35:35 +0000
Subject: [PATCH] tensor_converter: add frame padding removal for GRAY8
 format

- gst_tensor_converter_video_stride: for GRAY8 check if padding is added to
  width, height or frame size; if detected, set self->remove_padding flag
- gst_tensor_converter_chain: for video media type, when self->remove_padding
  is set, calculate frame padding and skip it when copying data to destination

Upstream-Status: Inappropriate [i.MX specific]

Signed-off-by: Alexandru Firuti <alexandru.firuti@nxp.com>
---
 gst/nnstreamer/elements/gsttensor_converter.c | 25 ++++++++++++++++---
 1 file changed, 22 insertions(+), 3 deletions(-)

diff --git a/gst/nnstreamer/elements/gsttensor_converter.c b/gst/nnstreamer/elements/gsttensor_converter.c
index 78f62247..3059e661 100644
--- a/gst/nnstreamer/elements/gsttensor_converter.c
+++ b/gst/nnstreamer/elements/gsttensor_converter.c
@@ -1032,7 +1032,7 @@ gst_tensor_converter_chain (GstPad * pad, GstObject * parent, GstBuffer * buf)
   GstTensorsConfig new_config;
   GstTensorInfo *_info;
   GstBuffer *inbuf;
-  gsize buf_size, frame_size;
+  gsize buf_size, frame_size, frame_padding;
   guint frames_in, frames_out;
   UNUSED (pad);
 
@@ -1111,12 +1111,22 @@ gst_tensor_converter_chain (GstPad * pad, GstObject * parent, GstBuffer * buf)
           }
         }
 
+        /* Calculate frame padding (padding at the end of entire frame) */
+        frame_padding = 0;
+       if (buf_size > frame_size) {
+           frame_padding = buf_size - frame_size;
+        }
+
         for (d0 = 0; d0 < frames_in; d0++) {
           for (d1 = 0; d1 < height; d1++) {
             memcpy (dest_info.data + dest_idx, src_info.data + src_idx, size);
             dest_idx += size;
             src_idx += offset;
           }
+         /* Skip frame padding if it exists (after all rows of current frame) */
+          if (frame_padding > 0 && d0 < (frames_in - 1)) {
+            src_idx += frame_padding;
+          }
         }
 
         gst_buffer_unmap (buf, &src_info);
@@ -1412,7 +1422,7 @@ gst_tensor_converter_get_format_list (GValue * list, ...)
  * @return TRUE if we need to add (or remove) stride per row from the stream data.
  */
 static gboolean
-gst_tensor_converter_video_stride (GstVideoFormat format, gint width)
+gst_tensor_converter_video_stride (GstVideoFormat format, gint width, gint height, gsize frame_size)
 {
   /**
    * @todo The actual list is much longer, fill them.
@@ -1420,6 +1430,10 @@ gst_tensor_converter_video_stride (GstVideoFormat format, gint width)
    */
   switch (format) {
     case GST_VIDEO_FORMAT_GRAY8:
+      if ((width % 16) || (height % 16) || (frame_size % 4096)) {
+        return TRUE;
+      }
+      break;
     case GST_VIDEO_FORMAT_RGB:
     case GST_VIDEO_FORMAT_BGR:
     case GST_VIDEO_FORMAT_I420:
@@ -1459,6 +1473,7 @@ gst_tensor_converter_parse_video (GstTensorConverter * self,
   GstVideoFormat format;
   gint width, height, views;
   guint i;
+  gsize size = 0;  // Frame size
 
   g_return_val_if_fail (config != NULL, FALSE);
 
@@ -1494,6 +1509,7 @@ gst_tensor_converter_parse_video (GstTensorConverter * self,
       config->info.info[0].dimension[0] = 1;
       config->info.info[0].dimension[1] = width;
       config->info.info[0].dimension[2] = height;
+      size = 1 * width * height;  // 1 byte per pixel
       break;
     case GST_VIDEO_FORMAT_GRAY16_BE:
     case GST_VIDEO_FORMAT_GRAY16_LE:
@@ -1501,6 +1517,7 @@ gst_tensor_converter_parse_video (GstTensorConverter * self,
       config->info.info[0].dimension[0] = 1;
       config->info.info[0].dimension[1] = width;
       config->info.info[0].dimension[2] = height;
+      size = 2 * 1 * width * height;  // 2 bytes per pixel
       break;
     case GST_VIDEO_FORMAT_RGB:
     case GST_VIDEO_FORMAT_BGR:
@@ -1508,6 +1525,7 @@ gst_tensor_converter_parse_video (GstTensorConverter * self,
       config->info.info[0].dimension[0] = 3;
       config->info.info[0].dimension[1] = width;
       config->info.info[0].dimension[2] = height;
+      size = 1 * 3 * width * height;  // 1 byte per channel, 3 channels
       break;
     case GST_VIDEO_FORMAT_RGBx:
     case GST_VIDEO_FORMAT_BGRx:
@@ -1521,6 +1539,7 @@ gst_tensor_converter_parse_video (GstTensorConverter * self,
       config->info.info[0].dimension[0] = 4;
       config->info.info[0].dimension[1] = width;
       config->info.info[0].dimension[2] = height;
+      size = 1 * 4 * width * height;  // 1 byte per channel, 4 channels
       break;
 #if GST_CHECK_VERSION(1, 20, 0)
     case GST_VIDEO_FORMAT_RGBP:
@@ -1550,7 +1569,7 @@ gst_tensor_converter_parse_video (GstTensorConverter * self,
    * Emit Warning if RSTRIDE = RU4 (3BPP) && Width % 4 > 0
    * @todo Add more conditions!
    */
-  if (gst_tensor_converter_video_stride (format, width)) {
+  if (gst_tensor_converter_video_stride (format, width, height, size)) {
     self->remove_padding = TRUE;
     silent_debug (self, "Set flag to remove padding, width = %d", width);
 
-- 
2.34.1

